packages/arb-validator/rollupvalidator/txtracker.go

```diff
@@ -110,214 +123,238 @@ func (a *assertionInfo) FindLogs(address *big.Int, topics []common.Hash) []logRe
+				logs = append(logs, logResponse{
+					Log:     evmLog,
+					TxIndex: txLogs.TxIndex,
+					TxHash:  txLogs.TxHash,
+				})
 			}
 		}
 	}
 	return logs
 }
 
-func newAssertionInfo() *assertionInfo {
-	return &assertionInfo{}
-}
-
+// txTracker is thread safe
 type txTracker struct {
-	txRequestIndex int
-	transactions   map[common.Hash]txInfo
-	assertionInfo  []*assertionInfo
-	assertionMap   map[common.Hash]*assertionInfo
-	accountNonces  map[common.Address]uint64
-	vmID           common.Address
-	requests       chan validatorRequest
+	rollup.NoopListener
+	chainAddress common.Address
+
+	// The RWMutex protects the variables listed below it
+	sync.RWMutex
+	txDB          *txDB
+	maxNodeHeight uint64
+	initialized   bool
 }
 
 func newTxTracker(
-	vmID common.Address,
-) *txTracker {
-	requests := make(chan validatorRequest, 100)
-	return &txTracker{
-		txRequestIndex: 0,
-		transactions:   make(map[common.Hash]txInfo),
-		assertionInfo:  make([]*assertionInfo, 0),
-		assertionMap:   make(map[common.Hash]*assertionInfo),
-		accountNonces:  make(map[common.Address]uint64),
-		vmID:           vmID,
-		requests:       requests,
+	db machine.CheckpointStorage,
+	ns machine.NodeStore,
+	chainAddress common.Address,
+) (*txTracker, error) {
+	txdb, err := newTxDB(db, ns, chainAddress)
+	if err != nil {
+		return nil, err
 	}
+	return &txTracker{
+		txDB:          txdb,
+		chainAddress:  chainAddress,
+		maxNodeHeight: 0,
+		initialized:   false,
+	}, nil
 }
 
-func (tr *txTracker) AssertionCount() <-chan int {
-	req := make(chan int, 1)
-	tr.requests <- assertionCountRequest{req}
-	return req
+// Delete assertion and transaction data from the reorged blocks if there are any
+func (tr *txTracker) RestartingFromLatestValid(_ context.Context, _ *rollup.ChainObserver, node *structures.Node) {
+	startDepth := node.Depth()
+	go func() {
+		tr.Lock()
+		defer tr.Unlock()
+		// First remove any data from reorged nodes
+		for i := tr.maxNodeHeight; i > startDepth; i-- {
+			if err := tr.txDB.removeUnconfirmedNode(i); err != nil {
+				continue
+			}
+		}
+		tr.maxNodeHeight = startDepth
+	}()
 }
 
-func (tr *txTracker) OutputMsgVal(assertionHash common.Hash, msgIndex int64) <-chan value.Value {
-	req := make(chan value.Value, 1)
-	tr.requests <- outputMsgRequest{assertionHash, msgIndex, req}
-	return req
+// AddedToChain is called when this listener is initially added to the
+// chain. If the listener was already added to a previous chain observer, we
+// must be restarting after a reorg and this function does nothing. When this
+// method is called for the first time, it processes all nodes that are valid,
+// but have not yet been confirmed and saved into the longterm db
+func (tr *txTracker) AddedToChain(_ context.Context, chain *rollup.ChainObserver) {
+	tr.Lock()
+	defer tr.Unlock()
+	if tr.initialized {
+		return
+	}
+	tr.initialized = true
+	nodesToProcess := chain.PendingCorrectNodes()
+	go func() {
```

Comment:
It might be better to acquire tr.Lock before launching the thread, then to defer tr.Unlock inside the thread. Presumably the purpose of launching a thread here is to avoid blocking the thread that is notifying us (so it can notify other components promptly).  But the current locking strategy risks having other things happen in the txtracker before the thread can run, which seems potentially risky.  Grabbing the lock before launching the thread, then releasing it inside the thread, will maintain continuous locking on the txtracker without holding up the caller.

Reply:
This pattern occurs at several places in this file. I assume the right answer (to change this, or not) will be the same in all of those places as it is here.

Reply:
I agree that this could be a potential source of problems. I just pushed a commit to switch this over to a more conservative locking strategy in all of the listener methods